metric_test.go

```diff
@@ -0,0 +1,223 @@
+			tok:   tokenMetric,
+			lit:   "testing",
+		},
+		{
+			input: "t|esting|ST[blah:blah]",
+			tok:   tokenMetric,
+			lit:   "t|esting",
+		},
+		{
+			input: "t|Sesting|ST[blah:blah]",
+			tok:   tokenMetric,
+			lit:   "t|Sesting",
+		},
+		{
+			input: "tel|a|M|Ssting|ST[blah:blah]",
+			tok:   tokenMetric,
+			lit:   "tel|a|M|Ssting",
+		},
+		{
+			input: "testing|ST[blah:blah]",
+			tok:   tokenMetric,
+			lit:   "testing",
+		},
+	}
+
+	for _, c := range cases {
+		buf := bytes.NewBufferString(c.input)
+		s := newMetricScanner(buf)
+		tok, lit, err := s.scanMetricName()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if tok != c.tok {
+			t.Error("failed to find the metric ident")
+		}
+		if lit != c.lit {
+			t.Error("incorrect literal scanned: ", c.lit, lit)
+		}
+	}
+}
+
+func TestScanTagSep(t *testing.T) {
+	stBuf := bytes.NewBufferString("|ST[blah:blah]")
+	s := newMetricScanner(stBuf)
+	tok, lit, err := s.peekTagSep()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if tok != tokenStreamTag {
+		t.Error("expected stream tag separator")
+	}
+
+	if lit != "|ST" {
+		t.Error("incorrect literal scanned: ", lit)
+	}
+
+	tok, lit, err = s.scanTagSep()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if tok != tokenStreamTag {
+		t.Error("expected stream tag separator")
+	}
+
+	if lit != "|ST" {
+		t.Error("incorrect literal scanned: ", lit)
+	}
+
+	mtBuf := bytes.NewBufferString("|MT[blah:blah]")
+	s = newMetricScanner(mtBuf)
+	tok, lit, err = s.peekTagSep()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if tok != tokenMeasurementTag {
+		t.Error("expected measurement tag separator")
+	}
+
+	if lit != "|MT" {
+		t.Error("incorrect literal scanned: ", lit)
+	}
+
+	tok, lit, err = s.scanTagSep()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if tok != tokenMeasurementTag {
+		t.Error("expected measurement tag separator")
+	}
+
+	if lit != "|MT" {
+		t.Error("incorrect literal scanned: ", lit)
+	}
+}
+
+func TestMetricParser(t *testing.T) {
+	var cases = []struct {
+		input string // input
+		num   int    // number of derived stream tags
+		lit   string // metric name literal
+	}{
+		{
+			input: "testing",
+			num:   0,
+			lit:   "testing",
+		},
+		{
+			input: "testing*",
+			num:   0,
+			lit:   "testing*",
+		},
+		{
+			input: "testing|ST[blah:blah]",
+			num:   1,
+			lit:   "testing",
+		},
+		{
+			input: "t|esting|ST[blah:blah]",
+			num:   1,
+			lit:   "t|esting",
+		},
+		{
+			input: "t|Sesting|ST[blah:blah]",
+			num:   1,
+			lit:   "t|Sesting",
+		},
+		{
+			input: "tel|a|M|Ssting|ST[blah:blah]",
+			num:   1,
+			lit:   "tel|a|M|Ssting",
+		},
+		{
+			input: "testing|ST[blah:blah]",
+			num:   1,
+			lit:   "testing",
+		},
+		{
+			input: "testing|ST[blah:blah,blah1:blah,blah2:blah]",
+			num:   3,
+			lit:   "testing",
+		},
+		{
+			input: `testing|ST[b"QUFB":blah,blah1:b"QkJCQgo=",b"YWFh":b"YmJi"]`,
+			num:   3,
+			lit:   "testing",
+		},
+		{
+			input: "testing|ST[blah:blah:blah:blah]",
+			num:   1,
+			lit:   "testing",
+		},
```

Comment:
The test suite for this only tests the metric names themselves, and not that the tag parsing is also working correctly.   There are also no tests for measurement tags, or names in a valid, but non-canonical form.  We also now have support for quoted strings in tags.   We likely want to add these to the test suite.   examples:

`foo|ST[a:b]|ST[b:c]` multiple `|ST[]` clauses
`foo|MT[a:b]` measurement tags
`foo|MT[a:b]|ST[b:d,e:f]|MT[g:h]|ST[i:j]` multiples of both
`foo|ST["quoted key string":"quoted value string"]`

https://docs.circonus.com/irondb/metric-names-tags-queries/ goes into more detail on what is valid or not.

Reply:
There was not yet support for measurement tags or multiple tag sets or quoted key and value strings in this parser. All of this functionality has now been added, and these unit tests have been added to the test suite.